bot/exts/moderation/voice_gate.py

```diff
@@ -160,6 +187,47 @@ async def on_message(self, message: discord.Message) -> None:
         with suppress(discord.NotFound):
             await message.delete()
 
+    @Cog.listener()
+    async def on_voice_state_update(
+            self,
+            member: Member,
+            before: discord.VoiceState,
+            after: discord.VoiceState
+    ) -> None:
+        """Pings a user if they've never joined the voice chat before and aren't voice verified."""
+        if member.bot:
+            log.trace("User is a bot. Ignore.")
+            return
+
+        # member.voice will return None if the user is not in a voice channel
+        if member.voice is None:
+            log.trace("User not in a voice channel. Ignore.")
+            return
+
+        in_cache = await self.redis_cache.get(member.id, NO_MSG)
+        if in_cache:
+            log.trace("User already in cache. Ignore.")
+            return
```

Comment:
I think this still doesn't behave properly. Having a default of `NO_MSG` is bad because then you cannot tell whether the member was or was not in the cache. I originally proposed that it checks whether `in_cache` is None (with a None default), but now I realize that it could also use `redis_cache.contains`, that would read the best.

Reply:
To expand on this: we want the if-condition to hold only if the member ID is not in the cache. If we `get` with a None default, `in_cache` being None means that the member wasn't there, and `NO_MSG` means that it *was* there but the message is deleted. But we need to check for the None explicitly, because both are falsey. If we use a `NO_MSG` default we cannot tell whether the ID was there or not, so it doesn't make sense for a default value.

Ultimately this should probably just be:
```python
if not await self.redis_cache.contains(member.id):
```

It did not occur to me initially, sorry.

Reply:
Resolved by 59b34c7.